Avoid removing effectful expressions in Closure, and eliminate more non-effectful ones.

alainfrisch opened the pull request:
When a function that ignores some of its parameters is inlined, the
corresponding arguments are still evaluated if they are not determined
to be "pure".  This commit improves this purity criterion by reusing
an existing function that detects more cases.  Only the non flambda
pipeline is adapted by this commit, but I guess the same optimization
is already part of flambda (CI will tell!) or would be easily added.

This optimization is triggered on the example discussed on the
caml-list in Dec. 2016 ("Closing the performance gap to C").

[EDIT]  This PR evolved from adding a simple optimization to fixing a series of bugs introduced
by badly estimating the effectfulness of primitives in the Closure pass.  Fixing these bugs is arguably more important than the new optimization.

alainfrisch committed:
Eliminate more useless computations when inlining

When a function that ignores some of its parameters is inlined, the
corresponding arguments are still evaluated if they are not determined
to be "pure".  This commit improves this purity criterion by reusing
an existing function that detects more cases.  Only the non flambda
pipeline is adapted by this commit, but I guess the same optimization
is already part of flambda (CI will tell!) or would be easily added.

alainfrisch commented:
This might be buggy, since is_pure_clambda incorrectly considers Pdivint as pure while it can raise an exception (would need to check if the previous uses of is_pure_clambda are affected).

alainfrisch commented:
The bug is_pure_clambda is a real one.  Currently, the following does not raise when compiled with ocamlopt (and it should):

````ocaml
type t = {x: int; y:int}
let f x = {x; y = x/0}.x
let () = ignore (f 1)
````

alainfrisch committed:
Pdivint is not pure, it can raise an exception.

alainfrisch committed:
Typo.

alainfrisch committed:
Non regression test.

alainfrisch commented:
Fixed Closure.is_pure_clambda.

gasche commented:
Wouldn't it be interesting to avoid catch-all patterns on instructions or primitive in this function? It would force us to consider each case and could uncover other similar bugs today, or tomorrow when a new instruction is added.

alainfrisch commented:
@gasche Agreed.  Note that the dangerous catch-all is on primitives (they are assumed to be pure "by default"), not on expressions.  But yes, being explicit would be a good idea in both cases.

alainfrisch commented:
Btw, array accesses look a bit fishy as well.  At the level of Clambda, the checkbound is not yet explicit, and is_pure_clambda still assumes loads from array to be pure.

alainfrisch commented:
Indeed:

````ocaml
type t = {x: int; y:int}
let x = {x = 1; y = [||].(0)}.x
````

doesn't raise when compiled with ocamlopt.

alainfrisch commented:
Also suspicious: Plazyforce, Prevapply, Pdirapply, Pmodint, Pdivbint, Pmodbint, Pstring_load_\*, Pbigstring_load_\*.

alainfrisch commented:
On the other hand, I don't see why Pduprecord is considered non pure.  Does anyone know?

alainfrisch commented:
Closure.is_pure is also suspicious...

gasche commented:
Pduprecord, at least if it corresponds to `{x with ...}`, may either do a initialization of all fields (pure) or a memcopy followed by field-by-field updates. I suppose people had the latter implementation in mind when marking it impure -- you could argue that it only mutates uniquely-owned memory, but I am not sure how true that is (could GC effects share the value or something?).

alainfrisch commented:
Mmh, I don't see how the "memcopy + updates" version would break invariants expected from pure expressions (in this context, "pure" is used to avoid evaluating an expression when its result is not observed; not e.g. to change the evaluation order (which would require to mark all memory reads as also "impure").

alainfrisch committed:
Exhaustive matching on primitives to define if they are 'pure'.

alainfrisch commented:
Ok, I've committed a version with an exhaustive match on primitives.  I did not try to take the "unsafe" arguments into account.

mshinwell commented:
Can this code just use Semantics_of_primitives instead, which already has a carefully-checked exhaustive list (not that I promise it has no bugs)?

(We may need a flag to allow it to accept primitives that are illegal in Flambda mode, but that's easy.)

alainfrisch commented:
Yes, indeed!  I'm a bit reluctant to have Closure depends on middle-end/.  Can we merge this module into Lambda?  One could imagine that either Simplif or alternative backend a-la Bucklescript use the information as well.

mshinwell commented:
I think the middle-end dependency may be there already (for Debuginfo), so I'm not sure we should worry about that.  If it really must be moved I'd rather it stayed as its own file and was moved properly with git to avoid introducing any merge conflicts with other branches in progress.

alainfrisch commented:
@mshinwell 

Can you comment on:

````
  | Prevapply
  | Pdirapply
  | Psequand
  | Psequor ->
    Misc.fatal_errorf "The primitive %a should have been eliminated by the \
        [Closure_conversion] pass."
      Printlambda.primitive prim
````

?

Aren't Prevapply/Pdirapply instead eliminated by `Simplif` (also for the non-flambda pipeline)?

About Psequand/Pseqor: would you be ok to support them in Semantics_of_primitive (returning `No_effects, No_coeffects`) instead of failing as above?

alainfrisch commented:
> I think the middle-end dependency may be there already (for Debuginfo), so I'm not sure we should worry about that. 

The middle-end is not linked in the bytecode compiler, only in tools/objinfo.

alainfrisch committed:
Move Semantics_of_primitives to bytecomp/ and use it in closure.

alainfrisch committed:
Make Semantics_of_primitives.for_primitive total so that it can be used in more contexts.

alainfrisch commented:
Ok, implemented @mshinwell's suggestion of reusing Semantics_of_primitives (after moving it to bytecomp/).  I removed error cases in this function, so that it can be applied in more contexts.

mshinwell commented:
@alainfrisch Can you check that [Flambda_invariants] checks for absence of the primitives for which you have removed the error cases?  (And make it check if it doesn't already.)

alainfrisch commented:
@mshinwell This is the case for Psequand, Pseqor, Pdirapply, Prevapply, but not Ploc.    At least this is implemented in Flambda_invariants.primitive_invariants (but I cannot guarantee this function is applied).

Is it worth adding a case for Ploc?

mshinwell commented:
Yeah, I think we might as well.  The more checks the better...

alainfrisch committed:
Check that Ploc does not appear in flambda.

alainfrisch commented:
Ok, done.  (I'm not sure, but it seems Plazyforce is also expanded by translcore.  And contrary to the error message in flambda_invariants, it seems that Prevapply/Pdirapply are handled by Simplif.)

mshinwell commented:
Feel free to fix the message!

alainfrisch commented:
The message might be sign that flambda does some extra useless work; it would be good to check first.

mshinwell commented on testsuite/tests/float-unboxing/float_subst_boxed_number.ml:
```diff
@@ -139,6 +139,11 @@ let unbox_minor_words () =
   done
 
+let ignore_useless_args () =
+  let f x _y = int_of_float (cos x) in
```
I think `f` should have `[@@inlined always]` to ensure that we're testing what you wanted to test, no?

alainfrisch replied:
I've added `[@inlined always]` to the call site (the attribute on the definition would be `@@inline`, no?).

mshinwell replied:
Fine

mshinwell commented on bytecomp/semantics_of_primitives.ml:
```diff
@@ -142,14 +142,17 @@ let for_primitive (prim : Lambda.primitive) =
+      No_effects, No_coeffects
   | Prevapply
-  | Pdirapply
+  | Pdirapply ->
```
Please un-indent these three lines to follow the rest of the file...

alainfrisch replied:
The indentation is coherent with was ocp-indent does, considering the definitions in the toplevel .ocp-indent.  It is also coherent with the style used elsewhere in the compiler (except in newest flambda code, I guess), and I find it more readable than the style used elsewhere in this file.  Since the file moves to bytecomp, I'd rather apply a style similar to other files in this directory, i.e. apply ocp-indent on the whole file.  Would you oppose to that?

alainfrisch replied:
I've done it (cf last commit), let me know if you are ok.

mshinwell replied:
I will manage (no doubt it will cause a merge conflict somewhere though, but not to worry)

mshinwell commented on asmcomp/closure.ml:
```diff
@@ -689,11 +693,8 @@ let bind_params loc fpc params args body =
-  | Lprim(_, args,_) -> List.for_all is_pure args
-  | Levent(lam, _ev) -> is_pure lam
+  | Lprim(p, args,_) -> is_pure_prim p && List.for_all is_pure args
+  | Levent(lam, _ev) -> is_pure lam (* can this happen in native code? *)
```
I think `Levent` may be present in a very few cases on the gdb branch, but I would have to check.  I would just remove this comment---it probably makes things more robust to include this case correctly, anyway.

alainfrisch replied:
Ok.

mshinwell commented on asmcomp/closure.ml:
```diff
@@ -448,6 +451,7 @@ let simplif_prim_pure fpc p (args, approxs) dbg =
   | Pfield n, [ Uprim(Pmakeblock _, ul, _) ], [approx]
     when n < List.length ul ->
       (List.nth ul n, field_approx n approx)
+
```
Gratuitous whitespace change.

alainfrisch replied:
Thanks, fixed.

mshinwell commented on Changes:
```diff
@@ -15,6 +15,10 @@ Next version (4.05.0):
   and "0 mod <expr>" in the case when <expr> was a non-constant
   evaluating to zero (Mark Shinwell)
 
+- Eliminate more useless pure computations when inlining functions
```
Reference GPR#983

alainfrisch replied:
Ok, and moved to the Bug Fix section.

mshinwell commented:
@alainfrisch Please change the `Flambda_invariants` message for `Pdirapply` and `Prevapply` to reference `Simplif` instead of `Closure_conversion`; I think they are indeed handled there.  I've made some other comments.  OK after that.

alainfrisch committed:
Take review comments into account.

alainfrisch committed:
Take review comments into account.

alainfrisch committed:
Apply standard indentation (as defined by .ocp-indent) and minor whitespace change.

mshinwell commented:
@alainfrisch I'm not sure about the `Ploc` case actually.  Shouldn't that "have effects"?  It's semantics depends on it not being moved, I thought.  (If we can't pin the semantics down properly it should maybe be a fatal error instead.)

alainfrisch commented:
Ploc evaluates to a compile-time constant, no?  Which effect could it have, and why could it not be moved?

mshinwell commented:
Well it seems to me that either `Ploc` has been compiled out, and if we only expect `Semantics_of_primitives` to be used on code after that compilation, we should have it produce a fatal error if it encounters this primitive.

However (reasonably enough) it seemed you were in favour of having `Semantics_of_primitives` do the right thing for any `Lambda` code, not just on that subset that we happen to generate at the moment, and in that world if `Ploc` exists I thought it would need to be treated as effectful to reflect the fact that if it were moved, it would produce a different result.

alainfrisch commented:
> to reflect the fact that if it were moved, it would produce a different result.

I guess that by moving, you mean "changing its source location", while I was rather interpreting moving as changing its position in "the tree" (Parsetree, Typedtree, Lambda, etc).

Currently, Ploc is expanded during the Typedtree->Lambda translation, but it could easily be done later, since the Lprim node also keeps the source location.  "Moving" Ploc within Lambda is fine as long as you don't change the Location.t attached to the Lprim node.

mshinwell commented:
By moving I meant a code motion transformation.  However from what you say it sounds like it's ok to do that anyway, as long as you don't change the location.  I think this is fine then.
